### PR TITLE
Add missing features used in gluon's source mapping

### PR DESCRIPTION
**Pull request by Marwes**
PR name is a bit blatant but I have been updating gluon to use codemap instead and found a few things that were missing (compared to the similar structure previously used in gluon). I think most things should be pretty straightforward improvements though naming may need to be changed.

https://github.com/gluon-lang/gluon/pull/505

### CODE REVIEW

File: codespan/src/filemap.rs [hunk 1]
```diff
@@ -117,6 +145,22 @@ impl FileMap {
         self.span
     }
 
+    pub fn offset(&self, line: LineIndex, column: ColumnIndex) -> Option<ByteOffset> {
+        self.line_offset(line).ok().and_then(|mut offset| {
+            offset += ByteOffset::from(column.0 as i64);
```
Comment (brendanzab) [inline codespan/src/filemap.rs]: Should we have an error if this runs into the next line?

Comment (brendanzab) [reply codespan/src/filemap.rs]: Kind of weird it is not called `byte_offset`, but then it would be inconsistent with `line_offset` 🤔

File: codespan/src/filemap.rs [hunk 2]
```diff
@@ -15,6 +15,30 @@ pub enum FileName {
...
+    fn from(name: String) -> FileName {
+        FileName::virtual_(name)
+    }
+}
+
+impl From<&'static str> for FileName {
+    fn from(name: &'static str) -> FileName {
+        FileName::virtual_(name)
+    }
+}
```
Comment (brendanzab) [inline codespan/src/filemap.rs]: Awesome stuff. I wasn't sure about adding these, but I think it's fine.

File: codespan/src/filemap.rs [hunk 3]
```diff
@@ -68,6 +92,10 @@ pub struct FileMap {
 }
 
 impl FileMap {
+    pub fn anonymous(src: String) -> FileMap {
+        Self::new(FileName::virtual_(""), src, ByteIndex::from(1))
```
Comment (brendanzab) [inline codespan/src/filemap.rs]: When is this used?

File: codespan/src/filemap.rs (hunk 1 above)
Comment (Marwes) [reply codespan/src/filemap.rs]: > Should we have an error if this runs into the next line?

Probably should have an error for that and to capture the error from `line_offset` as well. It gets a bit verbose when having precise errors.

File: codespan/src/filemap.rs (hunk 2 above)
Comment (Marwes) [reply codespan/src/filemap.rs]: Yeah, I don't think they should be that surprising and it was a bit tedious to need to import `FileName` frequently when I really just want to give a name to files added with https://github.com/brendanzab/codespan/blob/ccff649894f5ba3ce8c1c4c823a9ca2798d90ec8/codespan/src/codemap.rs#L31

File: codespan/src/filemap.rs (hunk 3 above)
Comment (Marwes) [reply codespan/src/filemap.rs]: Just for tests in my case. Maybe we want to make it `fn new(name: FileName, src: String) -> FileMap` though. I think there is merit to be able to construct standalone `FileMap`s in any case.

Comment (brendanzab) [reply codespan/src/filemap.rs]: Hum - yeah, I think what I was trying to avoid was invalid spans being created, like codemap does. But maybe I could just give up on that. 🤔

Comment (Marwes) [reply codespan/src/filemap.rs]: It is still easy to create two codemaps and accidentally use spans from one in the other though. Not really a solvable problem unfortunately :(

File: codespan/src/span.rs [hunk 4]
```diff
@@ -9,7 +10,7 @@ pub struct Span<I> {
     end: I,
 }
 
-impl<I: Index> Span<I> {
+impl<I: Ord> Span<I> {
```
Comment (brendanzab) [inline codespan/src/span.rs]: Why this change?

Comment (brendanzab) [reply codespan/src/span.rs]: I guess I was trying to limit it to fixed locations, rather than potentially folks using it for offsets. But I can see you reasoning, so let's go with that.

Comment (Marwes) [reply codespan/src/span.rs]: Okay, I figured it was just out of convenience that all functions were bounded by `I: Index`.  I could implement `Index` for my `Location` type and restore this bound so that guarantee remains instead (though `start` and `end` ought to still stay unbounded)

Comment (brendanzab) [reply codespan/src/span.rs]: Yeah, have a look - it might be a pain to implement for `Location` - eg. you might need to then introduce a corresponding `LocationOffset` type. I understand if that's a bit much to ask for though!

**Comment by Marwes** [comment]
This should be good now from my side at least

File: codespan/src/filemap.rs (hunk 3 above)
Comment (brendanzab) [reply codespan/src/filemap.rs]: Could we perhaps remove this and just make the `new` constructor public?

Comment (Marwes) [reply codespan/src/filemap.rs]: Done. I am a bit worried about exposing `start: ByteIndex` as it is not obvious that `0` is is not a valid index.  I hid that part of the public API so that it is only possible to create standalone `FileMap`s without `CodeMap`.
